### PR TITLE
Add basic SAML tests for mapping users.

### PR DESCRIPTION
**Pull request by clokep**
This adds some basic tests for SAML. This specifically aims to partially test the `_map_saml_response_to_user` method.

### CODE REVIEW

**Comment by clokep** [comment]
This was originally part of #8774. Unfortunately this will fail since libxmlsec is missing from our Buildkite images.

**Comment by clokep** [comment]
I wonder if we need to skip these tests if the package isn't found properly? Packagers might be unhappy otherwise.

File: tests/handlers/test_saml.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        config = self.default_config()
+        config["public_baseurl"] = BASE_URL
+        saml_config = {
+            "sp_config": {"metadata": {}},
+            # Disable grandfathering.
+            "grandfathered_mxid_source_attribute": None,
+            "user_mapping_provider": {"module": __name__ + ".TestMappingProvider"},
+        }
+        config["saml2_config"] = saml_config
```
Comment (richvdh) [inline tests/handlers/test_saml.py:69]: the normal way to do this is to override `default_config` and let `setup_test_homeserver` call it.

File: tests/handlers/test_saml.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+        hs = self.setup_test_homeserver(
+            http_client=self.http_client,
+            proxied_http_client=self.http_client,
+            config=config,
+        )
+
+        self.handler = SamlHandler(hs)
```
Comment (richvdh) [inline tests/handlers/test_saml.py:77]: is using our own `SamlHandler` (as opposed to the singleton `hs.get_saml_handler()`) deliberate?

File: tests/handlers/test_saml.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    ):
+        localpart = saml_response.ava["username"] + (str(failures) if failures else "")
+        return {"mxid_localpart": localpart, "displayname": None}
+
+
+class SamlHandlerTestCase(HomeserverTestCase):
+    def make_homeserver(self, reactor, clock):
+
+        self.http_client = Mock(spec=["get_json"])
+        self.http_client.user_agent = "Synapse Test"
```
Comment (richvdh) [inline tests/handlers/test_saml.py:59]: is this ever actually called? we don't seem to set up any results for `get_json` so I suspect not, in which case can we omit it, at least until it's needed?

File: tests/handlers/test_oidc.py [hunk 4]
```diff
@@ -832,7 +832,7 @@ def test_map_userinfo_to_user_retries(self):
         # test_user is already taken, so test_user1 gets registered instead.
         self.assertEqual(mxid, "@test_user1:test")
 
-        # Register all of the potential users for a particular username.
+        # Register all of the potential usernames for a particular username.
```
Comment (richvdh) [inline tests/handlers/test_oidc.py:835]:
```suggestion
        # Register all of the potential mxids for a particular OIDC username.
```

(similarly for SAML)

**Review by richvdh — CHANGES_REQUESTED** [review]
a few nits

File: changelog.d/8800.misc [hunk 5]
```diff
@@ -0,0 +1 @@
+Add tests for SAML integration.
```
Comment (clokep) [inline changelog.d/8800.misc:1]:
```suggestion
 Add additional error checking for OpenID Connect and SAML mapping providers.
```

**Comment by clokep** [comment]
Thanks for the review! Most of the comments were also applicable to the OIDC tests (which the SAML tests heavily cribbed from). I applied the requested changes to both test files!

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**